daemon/apparmor_default.go

```diff
@@ -25,7 +33,18 @@ func DefaultApparmorProfile() string {
 }
 
 func ensureDefaultAppArmorProfile() error {
-	if apparmor.HostSupports() {
+	// Restore the removed apparmor_parser in containerd:
+	// https://github.com/containerd/containerd/commit/1acca8bba36e99684ee3489ea4a42609194ca6b9
+	// Fixes: https://github.com/moby/moby/issues/44900
+	checkAppArmorParser.Do(func() {
+		if _, err := os.Stat("/sbin/apparmor_parser"); err == nil {
+			hasAppArmorParser = true
+		} else {
+			logrus.Warn("AppArmor enabled on system but \"apparmor_parser\" binary is missing, so profile can't be loaded")
+		}
+	})
+
+	if apparmor.HostSupports() && hasAppArmorParser {
```

Comment:
We should move `apparmor.HostSupports()` inside the `sync.Once` as that is what the original logic did, and we additionally need to actually check apparmor is enabled in the kernel before claiming that it is in a log message.

Reply:
Slightly wondering if we should put the `log.Warn` outside of the sync.Once though; maybe we want it to log on every invocation instead of only once?

Reply:
I think we should just wrap twice in the `sync.Once`, that's the best way to emulate the existing behavior (only checking for the binary once at this call site) we have. Similarly, I think that it makes more sense to log inside the sync.Once, as:

1. the logging is net-new and 'good enough' to tide us over until a followup refactor
2. it allows for the most minimal change/lets us preserve the exact logic from containerd

Reply:
Ah right, thanks for catching this 😄 
As for logging - I think it makes sense to do it inside the `sync.Once`.
Doing it every time will simply spam the daemon log. And it could be a good thing as it definitely attracts more attention than one log line somewhere in the beginning.
However in this case it seems that in most cases where the cause of not having the `apparmor_parser` is just the characteristic of the environment rather than user's error - and he can't do anything about it.